prov/mrail/src/mrail_cq.c

```diff
@@ -164,17 +161,32 @@ static int mrail_process_ooo_recvs(struct mrail_ep *mrail_ep,
 				   struct mrail_peer_info *peer_info)
 {
 	struct mrail_ooo_recv *ooo_recv;
+	struct mrail_recv *recv;
 	int ret;
 
-	while ((ooo_recv = mrail_peer_get_next(peer_info))) {
+	mrail_ep->lock_acquire(&mrail_ep->util_ep.lock);
+	ooo_recv = mrail_get_next_recv(peer_info);
+	if (!ooo_recv) {
+		goto exit;
+	}
+	while (ooo_recv) {
 		FI_DBG(&mrail_prov, FI_LOG_CQ, "found ooo_recv seq=%d\n",
 				ooo_recv->seq_no);
-		ret = mrail_cq_process_comp_buf_recv(mrail_ep, &ooo_recv->comp);
-		if (ret)
-			return ret;
-		mrail_free_ooo_recv(mrail_ep, ooo_recv);
-	}
+		recv = mrail_match_recv(mrail_ep, &ooo_recv->comp);
```

Comment:
Why is it safe to drop the lock while still accessing oob_recv and recv?  Can another thread acquire the same objects and modify either?  Does mrail_match_recv remove recv so that only this thread has access to it?

Reply:
This is the information I was looking for.  I was just asking for clarification, so I didn't have to read all of the code.  :)